lib/linter.js

```diff
@@ -242,6 +232,46 @@ class Linter {
     };
   }
 
+  /**
+   * @param {Object} options
+   * @param {string} options.source - The source code to verify.
+   * @param {string} options.filePath
+   * @param {string} options.moduleId
+   * @param {Object[]} results - The lint messages.
+   *
+   * @returns {string} output - The fixed source.
+   */
+  _applyFixes(options, results) {
+    let fixableIssues = results.filter(r => r.isFixable);
+
+    // nothing to do, bail out
+    if (fixableIssues.length === 0) {
+      return options.source;
+    }
+
+    let currentSource = options.source;
+
+    let pending = this.statusForModule('pending', options.moduleId);
+    let ruleNames = new Set(fixableIssues.map(r => r.rule));
+    for (let ruleName of ruleNames) {
+      let rule = this._buildRule(
+        ruleName,
+        {
+          pending,
+          shouldFix: true,
+          filePath: options.filePath,
+          rawSource: currentSource,
+        },
+        () => {}
+      );
+
+      let { code } = transform(currentSource, () => rule.getVisitor());
```

Comment:
The order is based on the ordering of message output from `.verify`. Basically we do this:

* determine the rule names for all the errors that have `isFixable` set
* loop through those rule names and run the fixer
* each rule runs independently (this ensures that two fixers cannot muck with each others AST)
  * `currentSource` is parsed, then passed to the rules visitor
  * `currentSource` is updated from the AST returned from the visitor (aka it is printed back to source)

> when a transform is applied on a node, should we somehow re-parse / re-lint the modified AST to apply a second fixer to the correct AST?

I'm not 100% sure I understand the question. For each fixer we reparse and lint, then we have a overall loop of "fixing iterations" (IIRC it is set to 10 max). So for a given set of errors we would do:

* run `.verify` to gather the current issues to be fixed
* run fixers (see list just above for how that is done)
* check if the fixers modified the output source
* if they did, and we have looped less than 10 times, start back at the beginning

Basically, ending up with `verify` -> `applyFixers` -> `verify` -> `applyFixers` -> etc.... (up to 10 times). The idea here is that even if two fixers did have an interaction (e.g. they both wanted to operate on the same node) we'd run both again until they "settle" (or error out because they can't agree).